camel/functions/search_functions.py

```diff
@@ -286,24 +359,61 @@ def search_google_and_summarize(query: str) -> str:
     for item in responses:
         if "url" in item:
             url = item.get("url")
-            # Extract text
-            text = text_extract_from_web(str(url))
+            # Extract text from the web page
+            try:
+                extracted_text = text_extract_from_web(str(url))
```

Comment:
why we still extract text from url rather than use response directly from search api?

Reply:
Thanks for reaching out with your question. Here are some of my thoughts.
First off, while search APIs like Google's are super handy, they come with a catch - there are limits on how much you can use them. Once you hit that cap, you can't make more requests without shelling out some extra cash. This can be a bit of a roadblock if we're doing a lot of data crunching.

Also, the info you get back from these APIs is more like a teaser of the web pages. It's helpful, but not enough for a deep dive. We need the full picture to make accurate summaries, and grabbing the text directly from the URLs gives us that.

Plus, dealing with the web pages ourselves lets us get fancy with our text analysis and customization. It's a bit more work, but it gives us more control and flexibility, especially when we're trying to dodge those API limits and keep our content fresh.

Reply:
Thanks for the reply! I think if it's due to the request limitation we can set the url extraction as plan B. If you see much limitation from the api response then it doesn't make sense to still get all the response from api but not use it at all.

Reply:
@jjyaoao Thanks for the detailed clarification, I agree with you that we need to extract from url for more detailed information. But if we don't use the response from the api, why we still append all these responses. Can we clean or split the function related to get response from api to save the resource?

Reply:
@Wendong-Fan Totally get your point, and I think it's a smart move. Here's what I'm thinking for the adjustments:

1. I've added a parameter to control the number of results returned from the URL, with a default value of 10. This way, the caller can adjust the number of results based on their needs.
2. I've implemented a conditional logic for the Google API: if the user provides an API key, the search will use Google's results; otherwise, it defaults to using DuckDuckGo. This ensures that only one search engine's response is returned.

Reply:
@jjyaoao Hi, have you committed to this modification based on the suggestion? If yes, please click "Resolve conversation", thanks.